js/util/pipeline-fixed.js

```diff
@@ -0,0 +1,387 @@
+ *   controlled by the `FinalBlockId` field in at least the last Data packet.
+ *
+ * The following logic is implemented in PipelineFixed:
+ *
+ * 1. Express the first Interest to discover the version:
+ *
+ *    >> Interest: /<prefix>?MustBeFresh=true
+ *
+ * 2. Infer the latest version of the Data: <version> = Data.getName().get(-2)
+ *
+ * 3. If the segment number in the retrieved packet == 0, go to step 5.
+ *
+ * 4. Pipeline the Interets starting from segment 0:
+ *
+ *    >> Interest: /<prefix>/<version>/<segment=0>
+ *    >> Interest: /<prefix>/<version>/<segment=1>
+ *    ...
+ *    >> Interest: /<prefix>/<version>/<segment=this.windowSize-1>
+ *
+ * At any given time the number of on the fly Interests should be equal
+ * to this.windowSize.
+ *
+ * 5. Pipeline the Interests startging from segment 1.
+ *
+ *    >> Interest: /<prefix>/<version>/<segment=1>
+ *    >> Interest: /<prefix>/<version>/<segment=2>
+ *    ...
+ *    >> Interest: /<prefix>/<version>/<segment=this.windowSize>
+ *
+ * 6. Upon receving a valid Data back we pipeline an Interest for the
+ * next expected segment.
+ *
+ *    >> Interest: /<prefix>/<version>/<segment=(N+1))>
+ *
+ * We repeat step 6 until the retrieved Data does not have a FinalBlockId
+ * or the FinalBlockId != Data.getName().get(-1).
+ *
+ * 7. Call the onComplete callback with a Blob that concatenates the content
+ *    from all the segmented objects.
+ *
+ * If an error occurs during the fetching process, the onError callback is called
+ * with a proper error code.  The following errors are possible:
+ *
+ * - `INTEREST_TIMEOUT`: if any of the Interests times out (probably after a number of retries)
+ * - `DATA_HAS_NO_SEGMENT`: if any of the retrieved Data packets don't have a segment
+ *   as the last component of the name (not counting the implicit digest)
+ * - `SEGMENT_VERIFICATION_FAILED`: if any retrieved segment fails
+ *   the user-provided VerifySegment callback or KeyChain verifyData.
+ * - `IO_ERROR`: for I/O errors when sending an Interest.
+ * -`DUPLICATE_SEGMENT_RECEIVED`: if any duplicate segment is receveid during
+ *   fetching the data (we do not ask for a segment that is already retrieved,
+ *   again)
+ *
+ * In order to validate individual segments, a KeyChain needs to be supplied.
+ * If verifyData fails, the fetching process is aborted with
+ * SEGMENT_VERIFICATION_FAILED. If data validation is not required, pass null.
+ *
+ *
+ * This is a public constructor to create a new PipelineFixed.
+ * If validatorKeyChain is not null, use it and ignore verifySegment.
+ * @param {string} basePrefix This is the prefix of the data we want to retreive
+ * its segments (excluding <version> and <segment> components).
+ * @param {Face} face The segments will be fetched through this face.
+ * @param {KeyChain} validatorKeyChain If this is not null, use its verifyData
+ * instead of the verifySegment callback.
+ * @param {function} verifySegment When a Data packet is received this calls
```

Comment:
To be clear, are you saying to make `onData` a callback function which is passed to pipeline?

Reply:
Yes. The function should take a `Data` and return a `Promise<boolean>`.
Note: although `Promise` is ES6-only, ndn-js already contains a polyfill called `SyncPromise`.

Reply:
Due to Jeff's comment, I did not applied this change.

Reply:
If you don't want to use a `Promise`, use a callback:

    verifySegment(data, cb)

`cb` receives a single boolean argument indicating whether the Data passes verification.

Synchronize `verifySegment` function does not make sense because it cannot retrieve certification.
`validatorKeyChain` and `verifySegment` have the same purpose so there should be only one in the API.

Reply:
The verify operation is asynchronous. It may need to do a network operation to retrieve a certificate for the verification, so the verify method cannot return immediately with the result. This is why the Validator API takes successCallback and failureCallback, where one is called some time later after all the network and crypto operations. (Some crypto operations in the browser are asynchronous.)

Reply:
https://github.com/named-data/ndn-js/blob/master/js/security/v2/validator.js#L125

Reply:
Jeff, about this comment, do you think it is necessary to remove either `validatorKeyChain` or `verifySegment`??

Reply:
The API for SegmentFetcher was to use either validatorKeyChain or verifySegment, but not both. (The API originally only had verifySegment which was synchronous. It's only there for backwards compatibility.) Notice that the call to validatorKeyChain.verifyData uses asynchronous callbacks:
https://github.com/named-data/ndn-js/blob/master/js/util/segment-fetcher.js#L231

Reply:
This is the security v1 API. The security v2 API does not use the KeyChain to verify. Instead, it has the separate Validator class that I linked to above. As part of this pull request, are you also updating to security v2? If that is too much for one pull request, you can stay with the asynchronous validatorKeyChain.verifyData, and drop verifySegment (and update to security v2 later).

Reply:
So, for this patch, I removed `verifySegment` and kept `validatorKeyChain` to validate the Data packets (if it was not `null`).